cogent/isa/Subtyping.thy

```diff
@@ -138,74 +138,32 @@ lemma type_lub_type_glb_commut:
   "K \<turnstile> c \<leftarrow> a \<squnion> b \<Longrightarrow> K \<turnstile> c \<leftarrow> b \<squnion> a"
   "K \<turnstile> c \<leftarrow> a \<sqinter> b \<Longrightarrow> K \<turnstile> c \<leftarrow> b \<sqinter> a"
 proof (induct rule: type_lub_type_glb.inducts)
-  case (lub_trecord ts ts1 ts2 s s2 s1)
-(*
+  case (lub_trecord K ts ts1 ts2 s s1 s2)
   then show ?case
-  proof (intro type_lub_type_glb.intros)
-    fix n t t1 t2 b b1 b2
-    assume
-      "(n, t, b) \<in> set ts"
-      "(n, t1, b1) \<in> set ts1"
-      "(n, t2, b2) \<in> set ts2"
-    then show "t \<leftarrow> t2 \<squnion> t1 \<and> b = inf b2 b1"
-      using lub_trecord
-      by (auto simp add: inf_commute)
-  qed simp+
+    apply (intro type_lub_type_glb.intros)
+    using list_all3_comm2 list_all3_mono apply fastforce
+         apply (induct rule: list_all3_induct)
```

Comment:
Try to avoid doing inductions inside inductions. Such a technique is very rarely necessary. Instead, in this case, I would recommend using `list_all3_conv_all_nth` as a simp rule.

For example, this subgoal can be solved by `apply (clarsimp simp add: list_all3_conv_all_nth, metis sup_commute)` without resorting to induction.

Reply:
It's sort of annoying, the `sup_commute` is throwing a spanner into Isabelle's automation. If we didn't have to use it, we could solve the entire goal with `by (fastforce intro!: type_lub_type_glb.intros simp add: list_all3_conv_all_nth)`, but adding the commut rule makes that solution loop.